core/src/typecheck/mod.rs

```diff
@@ -1589,6 +1590,11 @@ fn check<L: Linearizer>(
             );
             unify(state, &ctxt, ty, ty_import).map_err(|err| err.into_typecheck_err(state, rt.pos))
         }
+        // Types in term position are converted to contracts, so convert it before type-checking
+        // it. Note that we'll throw away the converted contract here and then do the conversion
+        // again during evaluation. If we had a mutable pass instead (like `subst`) we could
+        // do this work once.
+        Term::Types(typ) => check(state, ctxt, lin, linearizer, &typ.contract()?, ty),
```

Comment:
I don't think this makes a lot of sense in the long term. It makes sense because this is the current behavior, but truth is, this is never going to be well-typed (beside very trivial cases like primitive types, maybe):

```nickel
nickel> (let x = Number -> Number in 5) : _
error: incompatible types
 = Expected an expression of type `_a -> _b`
 = Found an expression of type `Dyn`
 = These types are not compatible
```

This is turned into applications of builtin contracts, here `$arrow $number $number`, but `$arrow` isn't typed, so it has type `Dyn`. I can see several possibilities:

1. Simply walk the type, instead of checking it. This would act as if the type was somewhere in an annotation. The cons is that we are too accepting: something absurd such as `Number -> 5` now typechecks. 
2. Show an error saying that types in term position aren't supported in statically typed blocks (probably the easiest, but the more restrictive)
3. Accept types which don't have any contracts (flat types) in them. Indeed, any "pure" type is well-formed and should be accepted. The issue is that the type of the content of a flat type can't be expressed in Nickel's type system (it's a union of `{_ : _}` and `Dyn -> Label -> Dyn` more or less), so there's no chance we can typecheck it in all generality currently. We could still accept some special cases, maybe.

I think I would favor 2., with the possibility of going to 3. incrementally. We probably have to go there before the next release though, otherwise 2. alone would be a regression as currently this typechecks (although this is a dubious behavior to rely on, as it's highly dependent of internal implementation details such as is this primitive type converted to a contract expression that has an application in it or not):

```nickel
(let x = Number in 5) : _
```

Reply:
I wasn't super keen on leaving a task that needed to be done before release, so I took a stab at the least permissive version of part 3: accept if and only if there are no flat types. If I understood it correctly, this is more permissive than the current rules but possibly less permissive than the eventual rule so it should be ok to release.

Reply:
Yup, sounds good :+1: I think it's fine to postpone a task to an immediately follow-up PR, just as a way of easing review and splitting work (not very different from splitting into commits), and not some vague unspecified later. But yeah the change isn't too big, so it's fine as it is